tokio/src/sync/rwlock.rs

```diff
@@ -0,0 +1,244 @@
+    permit: Permit,
+    lock: &'a RwLock<T>,
+}
+
+/// RAII structure used to release the exclusive write access of a lock when
+/// dropped.
+///
+/// This structure is created by the [`write`] and method
+/// on [`RwLock`].
+///
+/// [`write`]: struct.RwLock.html#method.write
+/// [`RwLock`]: struct.RwLock.html
+#[derive(Debug)]
+pub struct RwLockWriteGuard<'a, T> {
+    permits: Vec<ReleasingPermit<'a, T>>,
+    lock: &'a RwLock<T>,
+}
+
+// Wrapper arround Permit that releases on Drop
+#[derive(Debug)]
+struct ReleasingPermit<'a, T>(Permit, &'a RwLock<T>);
+
+impl<'a, T> Drop for ReleasingPermit<'a, T> {
+    fn drop(&mut self) {
+        self.0.release(&self.1.s);
+    }
+}
+
+// As long as T: Send, it's fine to send and share Mutex<T> between threads.
+// If T was not Send, sending and sharing a Mutex<T> would be bad, since you can access T through
+// Mutex<T>.
+unsafe impl<T> Send for RwLock<T> where T: Send {}
+unsafe impl<T> Sync for RwLock<T> where T: Send {}
+unsafe impl<'a, T> Sync for RwLockReadGuard<'a, T> where T: Send + Sync {}
+unsafe impl<'a, T> Sync for RwLockWriteGuard<'a, T> where T: Send + Sync {}
+
+impl<T> RwLock<T> {
+    /// Creates a new instance of an `RwLock<T>` which is unlocked.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::RwLock;
+    ///
+    /// let lock = RwLock::new(5);
+    /// ```
+    pub fn new(value: T) -> RwLock<T> {
+        RwLock {
+            c: UnsafeCell::new(value),
+            s: Semaphore::new(MAX_READS),
+            w: Mutex::new(()),
+        }
+    }
+
+    /// Locks this rwlock with shared read access, blocking the current task
+    /// until it can be acquired.
+    ///
+    /// The calling task will be blocked until there are no more writers which
+    /// hold the lock. There may be other readers currently inside the lock when
+    /// this method returns.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::sync::Arc;
+    /// use tokio::sync::RwLock;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let lock = Arc::new(RwLock::new(1));
+    ///     let c_lock = lock.clone();
+    ///
+    ///     let n = lock.read().await;
+    ///     assert_eq!(*n, 1);
+    ///
+    ///     tokio::spawn(async move {
+    ///         let r = c_lock.read().await;
+    ///         assert_eq!(*r, 1);
+    ///     })
+    ///}
+    /// ```
+    pub async fn read(&self) -> RwLockReadGuard<'_, T> {
+        let mut permit = ReleasingPermit(Permit::new(), self);
+        poll_fn(|cx| permit.0.poll_acquire(cx, &self.s))
+            .await
+            .unwrap_or_else(|_| {
+                // The semaphore was closed. but, we never explicitly close it, and we have a
+                // handle to it through the Arc, which means that this can never happen.
+                unreachable!()
+            });
+        RwLockReadGuard {
+            lock: self,
+            permit: std::mem::replace(&mut permit.0, Permit::new()),
```

Comment:
Why not just move the whole `ReleasingPermit` into the `RwLockReadGuard`?

Reply:
That's what we do in `write`. It also means we can delete the `Drop` impl.

Reply:
yeah you are totally right, missed it
thanks!